Firestore/core/test/unit/local/leveldb_index_manager_test.cc

```diff
@@ -62,11 +71,1042 @@ class LevelDbIndexManagerTest : public ::testing::Test {
+  // Server SDK.
+  persistence->Run("TestAdvancedQueries", [&]() {
+    index_manager->Start();
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "null", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "int", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "float", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "string", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "multi", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "array", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "array", model::Segment::kDescending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "array", model::Segment::kContains));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "map", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "map.field", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "prefix", model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "prefix", model::Segment::kAscending, "suffix",
+                       model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "a", model::Segment::kAscending));
+    index_manager->AddFieldIndex(MakeFieldIndex("coll", "a",
+                                                model::Segment::kAscending, "b",
+                                                model::Segment::kAscending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "a", model::Segment::kDescending, "b",
+                       model::Segment::kAscending));
+    index_manager->AddFieldIndex(MakeFieldIndex("coll", "a",
+                                                model::Segment::kAscending, "b",
+                                                model::Segment::kDescending));
+    index_manager->AddFieldIndex(
+        MakeFieldIndex("coll", "a", model::Segment::kDescending, "b",
+                       model::Segment::kDescending));
+    index_manager->AddFieldIndex(MakeFieldIndex("coll", "b",
+                                                model::Segment::kAscending, "a",
+                                                model::Segment::kAscending));
+
+    std::vector<nanopb::Message<google_firestore_v1_Value>> data;
+    data.push_back(Map());
+    data.push_back(Map("int", 1, "array", Array(1, "foo")));
+    data.push_back(Map("array", Array(2, "foo")));
+    data.push_back(Map("int", 3, "array", Array(3, "foo")));
+    data.push_back(Map("array", "foo"));
+    data.push_back(Map("array", Array(1)));
+    data.push_back(Map("float", -0.0, "string", "a"));
```

Comment:
I have a feeling that you had to change the canonical id because the Maps here aren't sorted. Can we sort them? Either manually or via `Map`.
If we sort manually we should probably also assert that `Map`'s input is sorted (to prevent headache for other people in the future).